### ISSUE TITLE
support python 3

### ISSUE THREAD

**Issue by snarfed** [snarfed/granary#114]
i believe the long pole here is a few indirect dependencies via oauth-dropins that don't support python 3. google-api-python-client? others? need to audit and come up with the complete list.

**Comment by kylewm** [comment]
`google-api-python-client` claims to support python 3!

**Comment by snarfed** [comment]
ooh! now if only they'd cut a release. :P google/google-api-python-client#421

**Comment by snarfed** [comment]
https://caniusepython3.com/project/granary claims we're only blocked on [gdata](https://github.com/snarfed/gdata-python-client-1) and [python-tumblpy](https://github.com/michaelhelmick/python-tumblpy).

**Comment by snarfed** [comment]
...and [tumblpy actually claims it supports python 3](https://github.com/michaelhelmick/python-tumblpy#features)!

gdata definitely does not: google/gdata-python-client#29 (also google/gdata-python-client#47)

here's a gdata fork that supposedly does: https://github.com/dvska/gdata-python3

**Comment by snarfed** [comment]
also, app engine standard is still python 2 only (and will probably always be), so i'll need to maintain python 2 compatibility. [six](https://pypi.python.org/pypi/six) and [future](https://pypi.python.org/pypi/future) are the two libraries for that i'm familiar with. not sure which is better.

**Comment by snarfed** [comment]
hrm, mox is also python 2 only: glasser/pymox#45

here's a fork that attempts python 3 compatibility: https://github.com/e0ne/pymox/tree/python3

if i use six or future, i could theoretically ship python 3 support while keeping the tests python 2 only, but...no thanks.

**Comment by snarfed** [comment]
sounds like [future](https://pypi.python.org/pypi/future) is preferred to [six](https://pypi.python.org/pypi/six). requires more recent versions of python 2 and, but it's more featureful and results in less crufty code.

https://docs.python.org/3/howto/pyporting.html#drop-support-for-python-2-6-and-older
http://python-future.org/faq.html#what-is-the-relationship-between-future-and-six
https://stackoverflow.com/questions/42110826/writing-cross-compatible-python-2-3-difference-between-future-six-and-fut

**Comment by snarfed** [comment]
note to myself, how i set up the virtualenv to use my local source clone of python-future:

```sh
cd l
foreach d (_dummy_thread _markupbase _thread builtins copyreg future html http libfuturize libpasteurize past queue reprlib socketserver tkinter winreg xmlrpc)
  ln -s ~/src/python-future/src/$d
end
```

**Comment by snarfed** [comment]
getting close! the `python3` branches of granary, bridgy, oauth-dropins, and webutil are all fully migrated to 2/3 compatible and all unit tests pass.

**Comment by snarfed** [comment]
woo, done! and merged into master. will be released in v1.12.
